[3.x] Fix RichTextLabel auto-wrapping on CJK texts

timothyqiu opened the pull request:
This PR fixes #21482.

There is still 1 px difference between when they wrap the last character after this fix. This is becauses `Label` autowraps if the text chunk width >= frame width, while `RichTextLabel` autowraps if text chunk width > frame width. I think this is not related to the original issue and worth a dedicated PR if desired.

Complex Text Layout on the `master` branch is the proper way to solve the wrapping problem, but it seems that it won't be backported.

This PR also supersedes #45290. I also added a fix to solve the issue mentioned in https://github.com/godotengine/godot/pull/45290#issuecomment-762410458, allowing word breaking on CJK and non-CJK edge. I'll revert this additional fix if it should be a separate PR.

---

Before:  (there is unexpected linebreak after "Hopefully, ", and "should" wraps after "sh")
![before](https://user-images.githubusercontent.com/372476/121187636-dc365880-c89a-11eb-84f5-1fad36fbd8c9.png)

After:
![after](https://user-images.githubusercontent.com/372476/121187640-dd678580-c89a-11eb-83ec-e42e95757cdc.png)

timothyqiu committed:
Fix RichTextLabel auto-wrapping on CJK

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -410,6 +412,18 @@ int RichTextLabel::_process_line(ItemFrame *p_frame, const Vector2 &p_ofs, int &
 
+						// For info about the unicode range, see Label::regenerate_word_cache.
+						const CharType current = c[end];
+						const bool separatable = (current >= 0x2E08 && current <= 0xFAFF) || (current >= 0xFE30 && current <= 0xFE4F);
```
Range probably should start with `2E80` not `2E08` and seems to be too broad (includes Cyrillic, Devanagari extended and other scripts that should not be separable).

Something like `2E80 - 9FCB`, `AC00 - D7FB`, `F900 - FAD9`, `FE30 - FE4F`, `20000 - 2FA1D` (The last range won't work on Windows, but should be supported on any other OS).

timothyqiu replied:
Yeah, I noticed that. Should I change the range used in `Label` too?

akien-mga replied:
Yeah that would make sense.

timothyqiu commented:
@bruvzg I used a slightly different range than you suggested.

* Ends CJK Unified Ideographs at `9FFF` instead of `9FCB`, the latter is the end of assigned code points as of Unicode 5.2, but expanded to the end of the block in later versions.
* Added `F900 - FAFF`, the [CJK Compatibility Ideographs block](https://en.wikipedia.org/wiki/CJK_Compatibility_Ideographs)
* Added `30000 - 3134F`, the [CJK Unified Ideographs Extension G block](https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_Extension_G)
* Added `FF65 - FF9F` and `FFA0 - FFDC` for halfwidth form of katakana and compatibility jamo

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.3.3.
